### ISSUE TITLE
CMYK option does not apply to QR codes

### ISSUE THREAD

**Issue by nicorsm** [pdfme/pdfme#460]
### Describe the bug

_TL;DR: when rendering documents as CMYK with the `colorType: cmyk` option, the QR codes inside them are rendered as RGB._

I'm trying to export a series of QR codes by using a template containing placeholders for a QR code, which has been created with the Template Designer at pdfme.com starting from a CMYK file. I'm also setting the `colorType` option to `cmyk`.
Everything looks good, however, when opening the output PDF file in Illustrator, a popup opens saying that the document has mixed RGB and CMYK content. 
I dag deeper by opening the document in Acrobat Pro and running the printing Preflight tool, and I found that, while the whole content is correctly rendered as CMYK, the QR code is rendered as RGB. 

![Screenshot 2024-03-27 at 18 14 33](https://github.com/pdfme/pdfme/assets/2906234/54fb70c4-536f-4726-a082-7c35278bbe42)

### To Reproduce

Sample code can be found here, together with a sample `output.pdf` file and the JSON of the template: https://github.com/nicorsm/pdfme-cmyk-issue
After installing the dependencies, you can run `node index.js`.

### Expected behavior

I would expect the QR code to be rendered as CMYK, same as the whole document.

### Your Environment

```markdown
- pdfme package(@pdfme/generator or @pdfme/ui): @pdfme/common @pdfme/generator
- pdfme version: 3.4.3
- Operating system: macOS Sonoma 14.0
- Node.js version or Browser name & version: v20.10.0, Safari 17.0
```

### Your Error Log

```shell
No error log from pdfme. Error is shown in Adobe Acrobat (see above) and the output report of the Preflight operation is attached to the repo (`output_report.txt`)
```

### Additional context

This issue has to do with the introduction of `colorType` option described here: https://github.com/pdfme/pdfme/pull/436

**Comment by hand-dot** [comment]
Ref: https://github.com/pdfme/pdfme/pull/436

**Comment by peteward** [comment]
Whilst CMYK colours can be passed to bwip-js, apparently it converts to RGB anyway internally. This might be difficult to achieve...

**Comment by hand-dot** [comment]
@peteward  Thank you for investigating 

umm...
It might be possible to achieve this by adding a process to convert the image data from RGB to CMYK before embedding it into pdf-lib.

**Comment by peteward** [comment]
The first step would probably be to do the conversion to CMYK for barcode color values as input for bwip to confirm if this does or does not work, then decide if post-conversion of images is something we can or should do

**Comment by nicorsm** [comment]
Hi @hand-dot, right now I'm not that familiar with the codebase of this library and with React, so I believe it's better to assign the task to someone else. I'll try anyway to look into this issue in the upcoming weeks as I want to understand better how the library works.

**Comment by necessarylion** [comment]
Just suggestion.

I think a good approach will be to convert barcode into svg which support in `bwip-js` and draw in pdf using `drawSvg`. But, by looking at the code, it seem like `drawSvg` was [force to use rgb color](https://github.com/pdfme/pdf-lib/blob/master/src/api/svg.ts#L374) only. 

So,
 - By Modify `drawSvg` to support CMYK color in `@pdfme/pdf-lib` (may be by converting hex color to CMYK) 
 - And then modify barcode schema to use svg instead.
 
Which will support color type for both `svg` and `barcode/qrcode` schemas at the same time.

**Comment by hand-dot** [comment]
@necessarylion  I see. This looks promising idea.

**Comment by tschuehly** [comment]
I've run into this issue aswell

**Comment by superiority-claw-agent** [comment]
Partially addressed in [PDFweave PR #5](https://github.com/IDNTEQ/pdfweave/pull/5) — commit [`5d4d15fa`](https://github.com/IDNTEQ/pdfweave/commit/5d4d15fa).

Shipped: CMYK input strings (`c100m0y0k0` syntax) now parse correctly and reach bwip-js without rejection.

**Not shipped:** true CMYK colorspace in the embedded PNG. bwip-js outputs an sRGB raster regardless of the colorspace of input strings, and pdf-lib's `drawImage` doesn't accept a CMYK PNG without a colorspace conversion pass. Doing this properly would need either pdf-lib SVG `device-cmyk` support or a PNG → CMYK conversion pre-embed. Documented as a follow-up in the PR.

If you need true CMYK output for print-production today, the workaround is to render the barcode as SVG separately and place the CMYK SVG yourself.

Context: pdfme/pdfme#1468.

— filed by @superiority-claw-agent on behalf of the PDFweave maintainers (IDNTEQ)

**Comment by hand-dot** [comment]
Confirmed with the published 6.1.12 packages: generating a QR code with `options.colorType: 'cmyk'` still embeds a three-component RGB image in the PDF. We checked the embedded image's color space directly, rather than judging its appearance on screen.

The print-color limitation remains. The earlier vector/SVG suggestion is a possible direction, but a fix must also verify the resulting PDF color space; switching representation alone would not establish CMYK output.

**Comment by hand-dot** [comment]
Maintainer triage follow-up on current `main` (`7501849ebe917e098e0a22569a1f338bf0fcc978`): **status: reproduces**.

Short rationale: the QR/barcode PDF path still renders the barcode through `bwip-js`, embeds it as a PNG with `pdfDoc.embedPng(...)`, and draws it with `page.drawImage(...)`. A runtime check generated a QR code with `options.colorType: 'cmyk'` and inspected the output PDF dictionaries; the embedded QR image still has `/ColorSpace /DeviceRGB`.

Suggested action: **keep open**. A fix still needs to make the barcode output itself CMYK-capable, or use another representation/path that can be verified as CMYK in the resulting PDF.

Evidence: I recorded the repro/check in the triage artifacts branch (not attached directly to this issue):
- Video: https://github.com/pdfme/pdfme/blob/cursor/batch-a-triage-9610/artifacts/batch-a/evidence-460-cmyk-qr.mp4
- Generated PDF: https://github.com/pdfme/pdfme/blob/cursor/batch-a-triage-9610/artifacts/batch-a/issue-460-cmyk-qr.pdf
- Notes: https://github.com/pdfme/pdfme/blob/cursor/batch-a-triage-9610/triage-460-cmyk-qr.md

The video shows the runtime command result and the key evidence: `embedded XObject color spaces: [{"name":"raw","colorSpace":"/DeviceRGB"}]`.

**Comment by hand-dot** [comment]
Maintainer triage follow-up on current `main` (`7501849ebe917e098e0a22569a1f338bf0fcc978`): **status: reproduces**.

Short rationale: the QR/barcode PDF path still renders the barcode through `bwip-js`, embeds it as a PNG with `pdfDoc.embedPng(...)`, and draws it with `page.drawImage(...)`. A runtime check generated a QR code with `options.colorType: 'cmyk'` and inspected the output PDF dictionaries; the embedded QR image still has `/ColorSpace /DeviceRGB`.

Suggested action: **keep open**. A fix still needs to make the barcode output itself CMYK-capable, or use another representation/path that can be verified as CMYK in the resulting PDF.

_(Removed an earlier mistaken evidence link; written triage findings kept.)_

**Comment by hand-dot** [comment]
Evidence correction: the linked MP4 evidence on the triage branch has been replaced with a real screen recording, not a generated summary slide. For this issue, `artifacts/batch-a/evidence-460-cmyk-qr.mp4` now shows an actual `xterm` session running `node artifacts/batch-a/repro-batch-a.mjs 460` and printing the `/DeviceRGB` QR image color-space result.

No issue status change: this still **reproduces** on current `main`.

**Comment by hand-dot** [comment]
**What this recording shows:** Repro: generate QR with CMYK options; embedded barcode image color space is still DeviceRGB (CMYK not applied).

Corrected evidence attachment: real screen recording of the terminal repro for #460. The video shows `node artifacts/batch-a/repro-batch-a.mjs 460` generating the QR PDF and printing the embedded image color space as `/DeviceRGB`.

Checked against `@pdfme/*` **6.1.13** (latest release) / `main` `7501849`.

![](https://github.com/user-attachments/assets/5b54c886-4362-407a-814d-7a7cef349c60)

**Comment by hand-dot** [comment]
**What this recording shows:** Terminal repro: generate QR with colorType/CMYK options; embedded QR image ColorSpace stays DeviceRGB.

Checked against `@pdfme/*` **6.1.13** / `main` `7501849`.

### Code run in the recording
```js
import { generate } from '@pdfme/generator';
import { BLANK_PDF } from '@pdfme/common';
import { barcodes, rectangle } from '@pdfme/schemas';
import { PDFDocument } from '@pdfme/pdf-lib';

const pdfBytes = await generate({
  template: {
    basePdf: BLANK_PDF,
    schemas: [[
      {
        name: 'qr',
        type: 'qrcode',
        position: { x: 20, y: 20 },
        width: 45,
        height: 45,
        barColor: '#000000',
        backgroundColor: '#ffffff',
      },
      {
        name: 'cmykRect',
        type: 'rectangle',
        position: { x: 80, y: 20 },
        width: 35,
        height: 35,
        color: '#00ffff',
      },
    ]],
  },
  inputs: [{ qr: 'https://pdfme.com', cmykRect: '' }],
  plugins: { qrcode: barcodes.qrcode, rectangle },
  options: { colorType: 'cmyk' },
});

const doc = await PDFDocument.load(pdfBytes);
// Inspect embedded QR image ColorSpace — expect /DeviceRGB (CMYK not applied to barcode)
```

![](https://github.com/user-attachments/assets/b71d4096-a052-41e4-860e-88ee52058399)
